proc_macros/src/lib.rs

```diff
@@ -0,0 +1,350 @@
+        }
+    }
+}
+
+fn get_from_usize_implementation(name: &Ident, variants: &[Ident]) -> proc_macro2::TokenStream {
+    quote! {
+        impl From<usize> for #name {
+            fn from(value: usize) -> Self {
+                match value {
+                    #(value if value == Self::#variants as usize => Self::#variants),*,
+                    _ => unreachable!(),
+                }
+            }
+        }
+    }
+}
+
+fn get_all_variants_collection_definition(
+    name: &Ident,
+    collection_type_name: &Ident,
+    variants: &[Ident],
+    all_variants_collection_name: &Ident,
+) -> proc_macro2::TokenStream {
+    quote! {
+        pub static #all_variants_collection_name: #collection_type_name<#name> = {
+            use SupportedLanguage::*;
+            BySupportedLanguage([
+                #(#variants),*
+            ])
+        };
+    }
+}
+
+fn get_collection_generator_macro_definition(
+    collection_type_name: &Ident,
+    variants: &[Ident],
+) -> proc_macro2::TokenStream {
+    let macro_name = format_ident!("{}", collection_type_name.to_string().to_snake_case());
+    quote! {
+        #[macro_export]
+        macro_rules! #macro_name {
+            ($($variant:ident => $value:expr),* $(,)?) => {
+                proc_macros::by_fixed_map!(
+                    [$($variant => $value),*],
+                    [
+                        #(#variants),*
+                    ],
+                    #collection_type_name
+                )
+            }
+        }
+    }
+}
+
+struct ByFixedMapArgs {
+    value_mapping: HashMap<Ident, Expr>,
+    variants: Vec<Ident>,
+    collection_type_name: Ident,
+}
+
+impl Parse for ByFixedMapArgs {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let value_mapping_content;
+        bracketed!(value_mapping_content in input);
+        let mut value_mapping: HashMap<Ident, Expr> = Default::default();
+        while !value_mapping_content.is_empty() {
+            let key: Ident = value_mapping_content.parse()?;
+            value_mapping_content.parse::<Token![=>]>()?;
+            let value: Expr = value_mapping_content.parse()?;
+            value_mapping.insert(key, value);
+            if value_mapping_content.is_empty() {
+                break;
+            }
+            value_mapping_content.parse::<Token![,]>()?;
+        }
+        input.parse::<Token![,]>()?;
+        let variants = parse_idents_array(input)?;
+        input.parse::<Token![,]>()?;
+        let collection_type_name: Ident = input.parse()?;
+        Ok(Self {
+            value_mapping,
+            variants,
+            collection_type_name,
+        })
+    }
+}
+
+#[proc_macro]
+pub fn by_fixed_map(input: TokenStream) -> TokenStream {
+    let ByFixedMapArgs {
+        value_mapping,
+        variants,
+        collection_type_name,
+    } = parse_macro_input!(input as ByFixedMapArgs);
+
+    if value_mapping.len() != variants.len() {
+        panic!("Incorrect variants");
```

Comment:
These could be more helpful and tell you eg which variant(s) were missing or what invalid variant names it encountered